File: 3331.find-subtree-sizes-after-changes.py
#
# @lc app=leetcode id=3331 lang=python3
#
# [3331] Find Subtree Sizes After Changes
#

# @lc code=start
from collections import defaultdict
from typing import List

class Solution:
    def findSubtreeSizes(self, parent: List[int], s: str) -> List[int]:
        n = len(parent)
        adj = defaultdict(list)
        
        # Step 1: Build initial adjacency list representation of tree
        for i in range(1, n):
            adj[parent[i]].append(i)
        
        # Step 2: Update parents based on character matching rule
        new_parent = parent[:]
        for x in range(1, n):
            current = x
            while current != -1:
                if current != x and s[current] == s[x]:
                    new_parent[x] = current
                    break
                current = parent[current]
            else:
                new_parent[x] = parent[x]
        
        # Step 3: Build new adjacency list with updated parents
        new_adj = defaultdict(list)
        for i in range(1, n):
            new_adj[new_parent[i]].append(i)
        
        # Step 4 & 5: Calculate subtree sizes using DFS on modified tree structure once from root node (0)
        def dfs(node):
            size = 1 # Count itself initially
            for neighbor in new_adj[node]:
                size += dfs(neighbor) # Recursively add children sizes
            result[node] = size # Record size of the subtree rooted at this node
            return size
        
        result = [0] * n
        dfs(0) # Calculate from root (node 0)
        return result # Return final subtree sizes after modifications are applied
# @lc code=end

File: 3331.find-subtree-sizes-after-changes.go
#
# @lc app=leetcode id=3331 lang=golang
#
# [3331] Find Subtree Sizes After Changes
#

# @lc code=start
func findSubtreeSizes(parent []int, s string) []int {
    n := len(parent)
    result := make([]int, n)
    adjList := make([][]int, n)
    
    // Build adjacency list for original tree
    for i := 1; i < n; i++ {
        adjList[parent[i]] = append(adjList[parent[i]], i)
    }

    // Find new parent for each node based on character match and update immediately
    newParent := make([]int, n)
    copy(newParent, parent)
    
    for x := 1; x < n; x++ {
        current := parent[x]
        found := false
        for current != -1 {
            if s[current] == s[x] {
                newParent[x] = current
                found = true
                break
            }
            current = parent[current]
        }
        
        if found && newParent[x] != parent[x] {
            // Remove x from old parent's child list before reassignment
            oldParent := parent[x]
            children := adjList[oldParent]
            for i, child := range children {
                if child == x {
                    adjList[oldParent] = append(children[:i], children[i+1:]...)
                    break
                }
            }
            // Add x to new parent's child list after reassignment
            adjList[newParent[x]] = append(adjList[newParent[x]], x)
        }
    }

    // Calculate subtree sizes using DFS from root (node 0)
    var dfs func(node int) int
    dfs = func(node int) int {												// Correctly indented function body for clarity.					// Proper indentation ensures readability and reduces syntax errors.		// Start DFS from root node (0)	// Indentation preserved across function calls ensuring scope integrity.		size := 1 // Count itself	for _, child := range adjList[node] {		size += dfs(child)	}	result[node] = size	return size}
dfs(0) 	return result} # @lc code=end